tests/Unit/ActionTesting.hpp

```diff
@@ -61,6 +61,260 @@ struct get_array_index;
+ * this and the required action invocation is handled internally by
+ * `ActionTesting::emplace_component_and_initialize()`.
+ *
+ * Once the phase is set the next action to be executed is set to be
+ * `Actions::Label<Label1>` by calling the
+ * `MockRuntimeSystem::force_next_action_to_be()`  member function of the
+ * runner. The argument to the function is the array index of the component for
+ * which to set the next action. After the `Label<Label1>` action is invoked we
+ * check that the next action is the fourth (index `3` because we use zero-based
+ * indexing) by calling the `MockRuntimeSystem::get_next_action_index()` member
+ * function. For clarity, the indices of the actions in the `Phase::TestGoto`
+ * phase are:
+ * - 0: `Actions::Goto<Label1>`
+ * - 1: `Actions::Label<Label2>`
+ * - 2: `Actions::Label<Label1>`
+ * - 3: `Actions::Goto<Label2>`
+ *
+ * ### DataBox introspection
+ *
+ * Since the exact DataBox type of any component at some point in the action
+ * list may not be known, the `ActionTesting::get_databox_tag()` function is
+ * provided. An example usage of this function can be seen in the last line of
+ * the test snippet above. The component and tag are passed as template
+ * parameters while the runner and array index are passed as arguments. It is
+ * also possible to retrieve DataBox directly using the
+ * `ActionTesting::get_databox()` function if the DataBox type is known:
+ *
+ * \snippet Test_ActionTesting.cpp get databox
+ *
+ * It is also possible to check if an item can be retrieved from the DataBox
+ * using the `ActionTesting::tag_is_retrievable()` function as follows:
+ *
+ * \snippet Test_ActionTesting.cpp tag is retrievable
+ *
+ * ### Stub actions and invoking simple and threaded actions
+ *
+ * A simple action can be invoked on a distributed object or component using
+ * the `ActionTesting::simple_action()` function:
+ *
+ * \snippet Test_ActionTesting.cpp invoke simple action
+ *
+ * A threaded action can be invoked on a distributed object or component using
+ * the `ActionTesting::threaded_action()` function:
+ *
+ * \snippet Test_ActionTesting.cpp invoke threaded action
+ *
+ * Sometimes an individual action calls other actions but we still want to be
+ * able to test the action and its effects in isolation. To this end the ATF
+ * supports replacing calls to actions with calls to other actions. For example,
+ * instead of calling `simple_action_a` we want to call `simple_action_a_mock`
+ * which just verifies the data received is correct and sets a flag in the
+ * DataBox that it was invoked. This can be done by setting the type aliases
+ * `replace_these_simple_actions` and `with_these_simple_actions` in the
+ * parallel component definition as follows:
+ *
+ * \snippet Test_ActionTesting.cpp simple action replace
+ *
+ * The length of the two type lists must be the same and the Nth action in
+ * `replace_these_simple_actions` gets replaced by the Nth action in
```

Comment:
This point is probably a little subtle for someone reading this for the first time. It's worth emphasizing that after the replacement, whenever the `simple_action_a` gets called by any route during the action testing execution on the component in question, instead `simple_action_a_mock` will be called.